File: utils/api_client.py
"""Reusable API client for REST API testing."""

import requests

from config.config import settings
from utils.logger import get_logger

logger = get_logger("api_client")


class APIClient:
    """HTTP client wrapper with logging, headers, and error handling."""

    def __init__(self, base_url: str | None = None):
        self.base_url = base_url or settings["api_base_url"]
        self.session = requests.Session()
        self.session.headers.update({
            "Content-Type": "application/json",
            "Accept": "application/json",
        })

    def get(self, endpoint: str, params: dict = None, **kwargs) -> requests.Response:
        url = f"{self.base_url}{endpoint}"
        logger.info("GET %s", url)
        response = self.session.get(url, params=params, **kwargs)
        logger.info("Response [%s]: %s", response.status_code, response.text[:200])
        return response

    def post(self, endpoint: str, json: dict = None, **kwargs) -> requests.Response:
        url = f"{self.base_url}{endpoint}"
        logger.info("POST %s | Body: %s", url, json)
        response = self.session.post(url, json=json, **kwargs)
        logger.info("Response [%s]: %s", response.status_code, response.text[:200])
        return response

    def put(self, endpoint: str, json: dict = None, **kwargs) -> requests.Response:
        url = f"{self.base_url}{endpoint}"
        logger.info("PUT %s | Body: %s", url, json)
        response = self.session.put(url, json=json, **kwargs)
        logger.info("Response [%s]: %s", response.status_code, response.text[:200])
        return response

    def patch(self, endpoint: str, json: dict = None, **kwargs) -> requests.Response:
        url = f"{self.base_url}{endpoint}"
        logger.info("PATCH %s | Body: %s", url, json)
        response = self.session.patch(url, json=json, **kwargs)
        logger.info("Response [%s]: %s", response.status_code, response.text[:200])
        return response

    def delete(self, endpoint: str, **kwargs) -> requests.Response:
        url = f"{self.base_url}{endpoint}"
        logger.info("DELETE %s", url)
        response = self.session.delete(url, **kwargs)
        logger.info("Response [%s]", response.status_code)
        return response

    def close(self):
        self.session.close()

File: src/api/ApiClient.ts
/**
 * API Client - Reusable HTTP client for API testing
 * Author: Shivam Sharma, Senior SDET
 *
 * Provides CRUD operations, auth flows, response-time assertions,
 * and schema validation using Ajv.
 */
import { APIRequestContext, APIResponse } from '@playwright/test';
import Ajv, { ValidateFunction } from 'ajv';

const ajv = new Ajv({ allErrors: true });

export interface ApiResponseWrapper {
  status: number;
  body: any;
  headers: Record<string, string>;
  responseTime: number;
}

export class ApiClient {
  private request: APIRequestContext;
  private baseUrl: string;
  private token?: string;

  constructor(request: APIRequestContext, baseUrl: string = 'https://jsonplaceholder.typicode.com') {
    this.request = request;
    this.baseUrl = baseUrl;
  }

  /** Set auth token for subsequent requests */
  setToken(token: string): void {
    this.token = token;
  }

  /** Build headers with optional auth */
  private getHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };
    if (this.token) {
      headers['Authorization'] = `Bearer ${this.token}`;
    }
    return headers;
  }

  /** Wrap API response with timing info */
  private async wrapResponse(response: APIResponse, startTime: number): Promise<ApiResponseWrapper> {
    const responseTime = Date.now() - startTime;
    let body: any;
    try {
      body = await response.json();
    } catch {
      body = await response.text();
    }
    const headers: Record<string, string> = {};
    const allHeaders = response.headers();
    for (const [key, value] of Object.entries(allHeaders)) {
      headers[key] = value;
    }
    return { status: response.status(), body, headers, responseTime };
  }

  /** GET request */
  async get(endpoint: string, params?: Record<string, string>): Promise<ApiResponseWrapper> {
    const url = params
      ? `${this.baseUrl}${endpoint}?${new URLSearchParams(params)}`
      : `${this.baseUrl}${endpoint}`;
    const startTime = Date.now();
    const response = await this.request.get(url, { headers: this.getHeaders() });
    return this.wrapResponse(response, startTime);
  }

  /** POST request */
  async post(endpoint: string, data: any): Promise<ApiResponseWrapper> {
    const startTime = Date.now();
    const response = await this.request.post(`${this.baseUrl}${endpoint}`, {
      data,
      headers: this.getHeaders(),
    });
    return this.wrapResponse(response, startTime);
  }

  /** PUT request */
  async put(endpoint: string, data: any): Promise<ApiResponseWrapper> {
    const startTime = Date.now();
    const response = await this.request.put(`${this.baseUrl}${endpoint}`, {
      data,
      headers: this.getHeaders(),
    });
    return this.wrapResponse(response, startTime);
  }

  /** PATCH request */
  async patch(endpoint: string, data: any): Promise<ApiResponseWrapper> {
    const startTime = Date.now();
    const response = await this.request.patch(`${this.baseUrl}${endpoint}`, {
      data,
      headers: this.getHeaders(),
    });
    return this.wrapResponse(response, startTime);
  }

  /** DELETE request */
  async delete(endpoint: string): Promise<ApiResponseWrapper> {
    const startTime = Date.now();
    const response = await this.request.delete(`${this.baseUrl}${endpoint}`, {
      headers: this.getHeaders(),
    });
    return this.wrapResponse(response, startTime);
  }

  /** Validate response body against JSON Schema */
  static validateSchema(data: any, schema: object): { valid: boolean; errors: string[] } {
    const validate: ValidateFunction = ajv.compile(schema);
    const valid = validate(data) as boolean;
    const errors = valid ? [] : (validate.errors || []).map(e => `${e.instancePath} ${e.message}`);
    return { valid, errors };
  }
}
